scripts/oneoff/export-approved-service-edits.py

```diff
@@ -0,0 +1,145 @@
+
+sys.path.insert(0, ".")
+
+from dmscripts.helpers.auth_helpers import get_auth_token
+
+
+ISO_FORMAT = "%Y-%m-%dT%H:%M:%S.%fZ"
+DEFAULT_DATETIME = "1970-01-01T00:00:00.000000Z"
+
+
+def parse_datetime(s: str) -> datetime:
+    try:
+        return datetime.strptime(s, ISO_FORMAT)
+    except ValueError as e:
+        if len(s) < len(DEFAULT_DATETIME):
+            return datetime.strptime(s + DEFAULT_DATETIME[len(s):], ISO_FORMAT)
+        else:
+            raise e
+
+
+def find_approved_service_edits(
+    data_api_client, from_datetime: datetime, to_datetime: Optional[datetime] = None
+) -> Iterator[dict]:
+    """Find service edits approved by users in a period"""
+
+    audit_events = data_api_client.find_audit_events_iter(
+        acknowledged="true", latest_first=True, audit_type=AuditTypes.update_service
+    )
+
+    # use datetimes to make our life easier
+    audit_events = map(
+        lambda e: {
+            **e,
+            "acknowledgedAt": datetime.strptime(e["acknowledgedAt"], ISO_FORMAT),
+        },
+        audit_events,
+    )
+
+    if to_datetime:
+        audit_events = filter(lambda e: e["acknowledgedAt"] < to_datetime, audit_events)
+
+    # Audit events can be acknowledged at any point after they are created, and
+    # because the server response is sorted by `createdAt` this means in theory
+    # we would need to search every page of the response to find all
+    # acknowledged in a time period. In practice service edits tend to get approved
+    # by admins in fairly regular blocks, so we just fudge it a bit.
+    # TODO: add an option to the api to return results sorted by `acknowledgedAt`
+
+    search_until: datetime = from_datetime - timedelta(
+        # anything more than 5 days takes a very long time
+        days=5
+    )
+    audit_events = itertools.takewhile(
+        lambda e: e["acknowledgedAt"] > search_until, audit_events
+    )
+    audit_events = filter(lambda e: from_datetime < e["acknowledgedAt"], audit_events)
+
+    # sort by acknowledgedAt, this can take a while
+    audit_events = sorted(audit_events, key=lambda e: e["acknowledgedAt"])
+
+    return audit_events
+
+
+if __name__ == "__main__":
+    args = docopt(__doc__)
+
+    stage = args["--stage"]
+
+    from_datetime: datetime = parse_datetime(args["<from>"])
+    to_datetime: Optional[datetime] = (
+        parse_datetime(args["<to>"]) if args["<to>"] else None
+    )
+
+    data_api_client = DataAPIClient(
+        base_url=get_api_endpoint_from_stage(stage),
+        auth_token=get_auth_token("api", stage),
+    )
+
+    audit_events = find_approved_service_edits(
+        data_api_client, from_datetime, to_datetime
+    )
+
+    # filter out acknowledgedBy which aren't email addresses (more likely to be scripts)
+    audit_events = filter(
+        lambda e: validate_email_address(e["acknowledgedBy"]), audit_events
+    )
+
+    # write details as csv
+    writer = csv.writer(sys.stdout)
```

Comment:
should this write to a file rather than `stdout`?

Reply:
Which would you prefer?

Reply:
@lfdebrux I don't mind, I would say whatever is consistent with other scripts that generate a CSV. But I don't have strong feelings about this.